File: booksamples/chapter-02/src/main/java/org/vulkanb/eng/Engine.java
package org.vulkanb.eng;

import org.vulkanb.eng.graph.Render;
import org.vulkanb.eng.scene.Scene;

public class Engine {

    private final IAppLogic appLogic;
    private final Render render;
    private final Scene scene;
    private final Window window;
    private boolean running;

    public Engine(String windowTitle, IAppLogic appLogic) {
        this.appLogic = appLogic;
        window = new Window(windowTitle);
        scene = new Scene(window);
        render = new Render(window, scene);
        appLogic.init(window, scene, render);
    }

    private void cleanup() {
        appLogic.cleanup();
        render.cleanup();
        window.cleanup();
    }

    public void run() {
        EngineProperties engineProperties = EngineProperties.getInstance();
        long initialTime = System.currentTimeMillis();
        float timeU = 1000.0f / engineProperties.getUps();
        double deltaUpdate = 0;

        long updateTime = initialTime;
        while (running && !window.shouldClose()) {
            window.pollEvents();

            long now = System.currentTimeMillis();
            deltaUpdate += (now - initialTime) / timeU;

            appLogic.input(window, scene, now - initialTime);

            if (deltaUpdate >= 1) {
                long diffTimeMilis = now - updateTime;
                appLogic.update(window, scene, diffTimeMilis);
                updateTime = now;
                deltaUpdate--;
            }

            render.render(window, scene);

            initialTime = now;
        }

        cleanup();
    }
    
    public void start() {
        running = true;
        run();
    }

    public void stop() {
        running = false;
    }
}

File: booksamples/chapter-05/kotlin/src/main/kotlin/org/vulkanb/eng/Engine.kt
package org.vulkanb.eng

import org.vulkanb.eng.graph.Render
import org.vulkanb.eng.scene.Scene

class Engine(windowTitle: String, private val appLogic: IAppLogic) {
    private val render: Render
    private val scene: Scene
    private val window: Window
    private var running = false

    init {
        window = Window(windowTitle)
        scene = Scene(window)
        render = Render(window, scene)
        appLogic.init(window, scene, render)
    }

    private fun cleanup() {
        appLogic.cleanup()
        render.cleanup()
        window.cleanup()
    }

    fun run() {
        var initialTime = System.currentTimeMillis()
        val timeU = 1000.0f / EngineProperties.ups
        var deltaUpdate = 0.0
        var updateTime = initialTime
        while (running && !window.shouldClose()) {
            window.pollEvents()
            val now = System.currentTimeMillis()
            deltaUpdate += ((now - initialTime) / timeU).toDouble()
            appLogic.input(window, scene, now - initialTime)
            if (deltaUpdate >= 1) {
                val diffTimeMilis = now - updateTime
                appLogic.update(window, scene, diffTimeMilis)
                updateTime = now
                deltaUpdate--
            }
            render.render(window, scene)
            initialTime = now
        }
        cleanup()
    }

    fun start() {
        running = true
        run()
    }

    fun stop() {
        running = false
    }
}
